src/numerics/petsc_matrix.C

```diff
@@ -1114,7 +1114,7 @@ void PetscMatrix<T>::add (const T a_in, SparseMatrix<T> & X_in)
   libmesh_assert_equal_to (this->n(), X_in.n());
 
   PetscScalar a = static_cast<PetscScalar>      (a_in);
-  PetscMatrix<T> * X = cast_ptr<PetscMatrix<T> *> (&X_in);
+  const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
```

Comment:
So, @jwpeterson's comment in #1915 about PETSc expecting a non-`const` pointer to a `Mat` isn't an issue here?

Reply:
Hmmm... my PETSc seemed happy with it but I suppose I should dig through old docs to make sure that wasn't a super recent change.

Reply:
Ah, I get it.  A Mat is a pointer-to-PETSc-stuff, and even if our PetscMatrix is const it can still give us a const-pointer-to-non-const-PETSc-stuff, so it doesn't matter whether or not PETSc's API here is philosophically const-correct (which is good, because it isn't).

Reply:
I've read this comment about 14 times and I'm still confused.

The change you're making on this line makes `X` a pointer-to-(const-PetscMatrix).  The thing you pass to PETSc is `X->_mat` which is of type `Mat`.  The PETSc API doesn't guarantee that `X->_mat` won't change during the call to `MatAXPY`.  So as far as I understand this should not compile.  What am I missing?

Reply:
C++ changes that X->_mat won't change during the call, because what MatAXPY gets is a *copy* of X->_mat.  Under the hood that Mat is itself a pointer to something else, which means the something else can change, even if (our copy of) the Mat can't.

Reply:
Oh, right.  Of course.  Nothing to see here.  Move along.